lib/seam/request.rb

```diff
@@ -72,6 +39,48 @@ def self.default_headers
         }
       end
 
+      class ResponseMiddleware < Faraday::Middleware
+        def initialize(app, options = {})
+          super(app)
+          @retry_options = options[:retry_options]
+        end
+
+        def on_complete(env)
+          return if env.success?
+
+          status_code = env.status
+          request_id = env.response_headers["seam-request-id"]
+
+          retry_statuses = @retry_options.fetch(:retry_statuses, [])
+          if retry_statuses.include?(status_code)
+            raise Faraday::RetriableResponse.new(nil, env.response)
+          end
+
+          raise Http::UnauthorizedError.new(request_id) if status_code == 401
+
+          body = begin
+            JSON.parse(env.body)
+          rescue
+            {}
+          end
```

Comment:
We can't assume the response is always JSON. It's possible for things in front of the API to return errors that are not JSON like Bad Gateway or Service Unavailable. I checked and the python SDK has the same assumption: https://github.com/seamapi/python/blob/7b62fa7eab7094d36af6f74465d4703da94d4330/seam/client.py#L64


We may need to replicate this logic in all the SDKs: https://github.com/seamapi/javascript-http/blob/8178e4f42e89198defdcda0a83a7b6ed4b482066/src/lib/seam/connect/error-interceptor.ts#L34-L61

I'm not sure we should keep the default "unknown" values either: the error needs to have a `type` and a `message` to be considered a Seam error. 

Basically:

1. Check for 401 status code and throw our custom SeamHttpUnauthorizedError
2. Try to determine if the error response can reasonably assumed to be a JSON error from the Seam API and throw it as either a `SeamHttpInvalidInputError` or the generic `SeamHttpApiError`. We can still default  
3. Otherwise, we just throw the error provided by the underlying http client.

Reply:
b35a6b17c89dd35e50de60682dd1de7d9c501457